common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java

```diff
@@ -22,49 +43,387 @@
+    mThrown.expectMessage("Peer indicated failure: Plain authentication failed: No authentication"
+        + " identity provided");
+    TTransport client = createClient(mTachyonConf, "", "correct-password");
+    try {
+      client.open();
+    } finally {
+      mServer.stop();
+    }
+  }
+
+  /**
+   * In CUSTOM mode, if client's password is empty, an exception should be thrown in server side.
+   */
+  @Test
+  public void customAuthenticationEmptyPasswordTest() throws Exception {
+    mTachyonConf.set(Constants.TACHYON_SECURITY_AUTHENTICATION, "CUSTOM");
+    mTachyonConf.set(Constants.TACHYON_AUTHENTICATION_PROVIDER_CUSTOM_CLASS,
+        MockAuthenticationProvider.class.getName());
+
+    // start server
+    startServerThread(mTachyonConf);
+
+    // check case that password is empty
+    mThrown.expect(TTransportException.class);
+    mThrown.expectMessage("Peer indicated failure: Plain authentication failed: No password "
+        + "provided");
+    TTransport client = createClient(mTachyonConf, "tachyon", "");
+    try {
+      client.open();
+    } finally {
+      mServer.stop();
+    }
+  }
+
+  /**
+   * TODO: In KERBEROS mode, ...
+   */
+  @Test
+  public void kerberosAuthenticationTest() throws Exception {
+    mTachyonConf.set(Constants.TACHYON_SECURITY_AUTHENTICATION, "KERBEROS");
+
+    // throw unsupported exception currently
+    mThrown.expect(UnsupportedOperationException.class);
+    mThrown.expectMessage("Kerberos is not supported currently.");
+    startServerThread(mTachyonConf);
+  }
+
+  private void startServerThread(TachyonConf conf) throws Exception {
+    // create args and use them to build a Thrift TServer
+    AuthenticationFactory factory = new AuthenticationFactory(conf);
+    TTransportFactory tTransportFactory = factory.getServerTransportFactory();
+
+    TServerSocket wrappedServerSocket = new TServerSocket(mServerAddress);
+
+    mServer = new TThreadPoolServer(new TThreadPoolServer.Args(wrappedServerSocket)
+        .maxWorkerThreads(2).minWorkerThreads(1)
+        .processor(null).transportFactory(tTransportFactory)
+        .protocolFactory(new TBinaryProtocol.Factory(true, true)));
+
+    // start the server in a new thread
+    Thread serverThread = new Thread(new Runnable() {
+      @Override
+      public void run() {
+        mServer.serve();
+      }
+    });
+
+    serverThread.start();
+
+    // ensure server is running, and break if it does not start serving in 2 seconds.
+    int count = 40;
+    while (!mServer.isServing() && serverThread.isAlive()) {
+      if (count <= 0) {
+        throw new RuntimeException("TThreadPoolServer does not start serving");
+      }
+      Thread.sleep(50);
+      count --;
+    }
+  }
+
+  /**
+   * This customized authentication provider is used in CUSTOM mode. It authenticate the user by
+   * verifying the specific username:password pair.
+   */
+  public static class MockAuthenticationProvider implements AuthenticationProvider {
```

Comment:
Because MockAuthenticationProvider strictly checks username and password, let's rename it, e.g. `exactAuthenticationProvider`?